core/server/proxy/src/main/java/alluxio/proxy/s3/signature/StringToSignProducer.java

```diff
@@ -0,0 +1,326 @@
+          60 * 60 * 24 * 7;
+
+  private StringToSignProducer() {
+  }
+
+  /**
+   * Convert signature info to strToSign.
+   *
+   * @param signatureInfo
+   * @param context
+   * @return signature string
+   * @throws Exception
+   */
+  public static String createSignatureBase(
+      SignatureInfo signatureInfo,
+      ContainerRequestContext context
+  ) throws Exception {
+    return createSignatureBase(signatureInfo,
+        context.getUriInfo().getRequestUri().getScheme(),
+        context.getMethod(),
+        context.getUriInfo().getRequestUri().getPath(),
+        LowerCaseKeyStringMap.fromHeaderMap(context.getHeaders()),
+        fromMultiValueToSingleValueMap(
+        context.getUriInfo().getQueryParameters()));
+  }
+
+  /**
+   * Convert request info to strToSign.
+   *
+   * Construct String to sign in below format.
+   * StringToSign =
+   * Algorithm + \n +
+   * RequestDateTime + \n +
+   * CredentialScope + \n +
+   * HashedCanonicalRequest
+   *
+   * For more details refer to AWS documentation:
+   * https://docs.aws.amazon.com/general/latest/gr/sigv4-create-string-to-sign.html
+   *
+   * @param signatureInfo
+   * @param scheme
+   * @param method
+   * @param uri
+   * @param headers
+   * @param queryParams
+   * @return signature string
+   * @throws Exception
+   */
+  public static String createSignatureBase(
+       SignatureInfo signatureInfo,
+       String scheme,
+       String method,
+       String uri,
+       LowerCaseKeyStringMap headers,
+       Map<String, String> queryParams
+  ) throws Exception {
+    StringBuilder strToSign = new StringBuilder();
+    String credentialScope = signatureInfo.getCredentialScope();
+
+    // If the absolute path is empty, use a forward slash (/)
+    uri = (uri.trim().length() > 0) ? uri : "/";
+    // Encode URI and preserve forward slashes
+    strToSign.append(signatureInfo.getAlgorithm() + NEWLINE);
+    strToSign.append(signatureInfo.getDateTime() + NEWLINE);
+    strToSign.append(credentialScope + NEWLINE);
+
+    String canonicalRequest = buildCanonicalRequest(
+         scheme,
+         method,
+         uri,
+         signatureInfo.getSignedHeaders(),
+         headers,
+         queryParams,
+         !signatureInfo.isSignPayload());
+    strToSign.append(hash(canonicalRequest));
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("canonicalRequest:[{}], StringToSign:[{}]", canonicalRequest, strToSign);
+    }
+
+    return strToSign.toString();
+  }
+
+  /**
+   * Convert MultivaluedMap to a single value map.
+   *
+   * @param queryParameters
+   * @return a single value map
+   */
+  public static Map<String, String> fromMultiValueToSingleValueMap(
+            MultivaluedMap<String, String> queryParameters) {
+    Map<String, String> result = new HashMap<>();
+    for (String key : queryParameters.keySet()) {
+      result.put(key, queryParameters.getFirst(key));
+    }
+    return result;
+  }
```

Comment:
See point 2 on my comment regarding `LowerCaseKeyStringMap`, but I don't think we need to handle the case where a client has provided duplicate query parameters. From my understanding S3 expects those to be provided at most once.